Upgrade mechanism (#23).

paulodamaso opened the pull request:
For #23: Created tests for `HtUpgradeWire` behavior.

paulodamaso commented:
@llorllale Could yout take a look at this, looks like Zerocrat is not assigning this `REV` task to anyone.

0crat commented:
> @0crat status ([here](https://github.com/yegor256/cactoos-http/issues/53#issuecomment-389216670))

@llorllale This is what I know about this job, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#53](https://github.com/yegor256/cactoos-http/issues/53) is [in scope](http://www.zerocracy.com/policy.html#14) for 3days
  * The role is `REV`
  * The job is not assigned to anyone
  * The [budget](http://www.zerocracy.com/policy.html#4) is not set yet
  * These users are banned and won't be assigned:
    * @paulodamaso[/z](https://www.0crat.com/u/paulodamaso): This user reported the ticket
  * Job [footprint](https://www.0crat.com/footprint/C63314D6Z?q=%7Bjob%3A%27gh%3Ayegor256%2Fcactoos-http%2353%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

<!-- https://www.0crat.com/footprint/C63314D6Z/1351579069, version: 1.0-SNAPSHOT, hash: 91fcb4fb -->

0crat commented:
This pull request [#53](https://github.com/yegor256/cactoos-http/issues/53) is assigned to @vssekorin[/z](https://www.0crat.com/u/vssekorin), here is [why](https://www.0crat.com/footprint/C63314D6Z/1361124312); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @llorllale[/z](https://www.0crat.com/u/llorllale) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer

<!-- https://www.0crat.com/footprint/C63314D6Z/1361125319, version: 1.0-SNAPSHOT, hash: 91fcb4fb -->

VsSekorin commented on src/test/java/org/cactoos/http/HtUpgradeWireTest.java:
```diff
@@ -0,0 +1,167 @@
+         * Returns the wire of this request.
+         * @return Wire.
+         */
+        public Wire wire() {
```
Public methods without `@Override`. You can `implements Input, Scalar<Wire>`.

VsSekorin commented on src/test/java/org/cactoos/http/HtUpgradeWireTest.java:
```diff
@@ -0,0 +1,167 @@
+        /**
+         * Origin response.
+         */
+        private final HtResponse origin;
```
Why is this called `origin`, and not a `response`? `origin` is fimilar to decorators, but `HtResponse` is not `Input`.

paulodamaso replied:
@VsSekorin Didn't unsderstood your remark, `HtResponse` is an `Input` indeed.

VsSekorin replied:
My bad. Thanks

paulodamaso commented:
@VsSekorin Implemented suggestions, thank you!

VsSekorin commented:
@rultor good to merge

rultor commented:
> @rultor good to merge

@vssekorin Thanks for your request. @llorllale Please confirm this.

paulodamaso commented:
@llorllale Please take a look.

llorllale commented on src/main/java/org/cactoos/http/HtUpgradeWire.java:
```diff
@@ -0,0 +1,59 @@
+ * @todo #23:30 min This implementation of class HtUpgradeWire is just
+ *  forwarding origin wire behavior. Finish implementation so a HtUpgradeWire
+ *  can be upgraded to a HtSecureWire on 101 - Swithicng Protocols status code,
+ *  switching the wire to https protocol. The test HtUpgradeWireTest#testUpgrade
```
@paulodamaso I don't understand why the test needs to be *removed*. Seems like a good outline for a perfectly valid test scenario.

llorllale replied:
@paulodamaso please fix the text on this puzzle accordingly

llorllale commented:
@paulodamaso @VsSekorin not trying to sound harsh, but I am having trouble understanding exactly how this PR moves us forward in the direction of the "upgrade mechanism" of this task, aside from introducing the class' name `HtUpgradeWire` and two tests (one of which apparently we need to remove?)

paulodamaso commented:
@llorllale Well, this PR does implements the test to guarantee that the 
we will have an `HtSecureWire` when we receive an `101 - Swithcing protocols`. I've faked an 101 code and I am expecting that `HtUpgradeWire` uses an `HtSecureWire`. I've started this PR with the tests of the wire upgrade feature, and it wasn't possible to implement `HtUpgradeWire` class in job timebox; so, I've implemented an dummy implementation of `HtUpgradeWire` just to model my tests, implemented the tests, and left a `@todo` about it.

llorllale commented:
@paulodamaso the outline of the `Upgrade` mechanism: the client initiates by asking for an upgrade in the request (our users will build this request) and if the server responds with the `Upgrade` header then the client needs to support the kind of upgrade specified in the response. Upgrades are typically done from HTTP/1.1 -> HTTP/2, although sometimes HTTP -> HTTPS is used as well.

`HtUpgradeWire` should therefore be configurable; you should be able to specify the wires that you, as the client, are able to upgrade to.

As a rough first draft, we could possibly be do something like:

```java
new HtUpgradeWire(
    new IterableOf<>(
        new MapEntry<Func<String, Boolean>, Func<URI, Wire>>(
            upgrade -> upgrade.contains("TLS"),
            HtSecureWire::new
        )
    )
)
```

paulodamaso commented:
@llorllale I've corrected `@todo` about implementation of  `HtUpgradeWire`. I think that the above comment should be added to the puzzle that will be created by this `@todo`.

llorllale commented:
@paulodamaso I'm afraid I can't accept just that. The job of presenting the research rests on the job's assignee. If you had done that then I would have been more willing to accept this PR. 

But now, since I presented the research, I expect more than just a puzzle with the research (unless you do some more research that uncovers *even more* stuff that is relevant to completing the task or that acts as an impediment in some way).

paulodamaso commented:
@llorllale Please take a look at my comment above, https://github.com/yegor256/cactoos-http/pull/53#issuecomment-391132150 . It explains what I've done in this PR, and why I'm suggesting this course of action. Notice that I'm not just asking a puzzle with the research, I've created the test that models the update wire class behavior according to the documentation provided at the time. I'm afraid that the research presented by you, while useful, is related to puzzle I've added to the code, that asks for `HtUpgradeWire` implementation according to the behavior defined by the test class implemented in this PR.
In my point of view this PR match what was asked in issue description (and trying to follow what said in https://www.yegor256.com/2015/01/15/how-to-cut-corners.html ) ; if you think that I still have to do something here, please clearly define what is it so I can do it and close the issue.

llorllale commented:
@paulodamaso thinking about it more, the "upgrade" is not the job of a `Wire`, because it would require reading the contents of the response. This is the job of an `HtResponse`-like object. A `Wire` is just a dumb pipe that sends and receives bytes.

So my draft code above should be refactored to:

```java
new HtUpgradedResponse(
    new IterableOf<>(
        new MapEntry<Func<String, Boolean>, Func<URI, Wire>>(
            upgrade -> upgrade.contains("TLS"),
            HtSecureWire::new
        )
    )
)
```

Fix the `@todo` with this and let's call it a day.

paulodamaso commented:
@llorllale Changes made, take a look.

rultor commented:
> @rultor merge

@llorllale Done! FYI, the full log is [here](http://www.rultor.com/t/14567-393172990) (took me 11min)

0crat commented:
The job [#53](https://github.com/yegor256/cactoos-http/issues/53) is now out of scope

<!-- https://www.0crat.com/footprint/C63314D6Z/1501446621, version: 1.0-SNAPSHOT, hash: 43f99e91 -->

elenavolokhova commented:
@VsSekorin According to our Policy:
> The code reviewer found at least three problems in the code

Only two issues were found in this review. Please, confirm that you will try to find at least three problems next time.

0crat commented:
Order was finished, quality is "acceptable": +15 point(s) just awarded to @vssekorin[/z](https://www.0crat.com/u/vssekorin)

<!-- https://www.0crat.com/footprint/C63314D6Z/1551150190, version: 0.22, hash: ${buildNumber} -->

0crat commented:
Quality review completed: +8 point(s) just awarded to @elenavolokhova[/z](https://www.0crat.com/u/elenavolokhova)

<!-- https://www.0crat.com/footprint/C63314D6Z/1551150235, version: 0.22, hash: ${buildNumber} -->
